cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
I don't think this is the correct fix. Looking further down at the rest of the code there's at least one place where we're not verifying that we're not at the end of the string. For example, if the string `"-"` were passed into this function the `while (*charp == '0') {...}` would dereference some memory region 1 byte (`sizeof(char)`) past the end of the iterator which would likely cause the program to abort.